File: packages/lmon/src/tokenizer.ts
export enum TokenType {
  LPAREN = 'LPAREN',
  RPAREN = 'RPAREN',
  LBRACE = 'LBRACE',
  RBRACE = 'RBRACE',
  LBRACKET = 'LBRACKET',
  RBRACKET = 'RBRACKET',
  COLON = 'COLON',
  COMMA = 'COMMA',
  LABEL_OR_VALUE = 'LABEL_OR_VALUE',
  QUOTED_STRING = 'QUOTED_STRING',
  ARRAY_MARKER = 'ARRAY_MARKER',
  NEWLINE = 'NEWLINE',
  EOF = 'EOF',
}

export interface Token {
  type: TokenType;
  value: string;
  line: number;
  column: number;
}

export function tokenize(input: string): Token[] {
  const tokens: Token[] = [];
  let i = 0;
  let line = 1;
  let column = 1;

  const current = (): string | undefined => input[i];
  const peek = (offset = 1): string | undefined => input[i + offset];
  const advance = (): void => {
    if (input[i] === '\n') {
      line++;
      column = 1;
    } else {
      column++;
    }
    i++;
  };

  const skipWhitespace = (): void => {
    while (current() && /[ \t\r]/.test(current()!)) {
      advance();
    }
  };

  const readQuotedString = (quote: string): string => {
    advance();
    let value = '';
    while (current() && current() !== quote) {
      if (current() === '\\') {
        advance();
        const escaped = current();
        if (escaped === 'n') value += '\n';
        else if (escaped === 't') value += '\t';
        else if (escaped === '\\') value += '\\';
        else if (escaped === quote) value += quote;
        else value += escaped;
        advance();
      } else {
        value += current();
        advance();
      }
    }
    if (current() === quote) advance();
    return value;
  };

  const readBareWord = (): string => {
    let value = '';
    while (current() && /[a-zA-Z0-9_]/.test(current()!)) {
      value += current();
      advance();
    }
    return value;
  };

  const readUnquotedValue = (): string => {
    let value = '';
    while (
      current() &&
      current() !== ',' &&
      current() !== '{' &&
      current() !== '}' &&
      current() !== '[' &&
      current() !== ']' &&
      current() !== '\n' &&
      /\S/.test(current()!)
    ) {
      if (current() === '\\') {
        advance();
        const escaped = current();
        if (escaped === 'n') value += '\n';
        else if (escaped === 't') value += '\t';
        else if (escaped === '\\') value += '\\';
        else if (escaped === ',') value += ',';
        else if (escaped === '{') value += '{';
        else if (escaped === '}') value += '}';
        else if (escaped === '[') value += '[';
        else if (escaped === ']') value += ']';
        else if (escaped === ':') value += ':';
        else value += escaped;
        advance();
      } else {
        value += current();
        advance();
      }
    }
    return value.trimEnd();
  };

  while (i < input.length) {
    skipWhitespace();

    if (i >= input.length) break;

    const ch = current();
    const startLine = line;
    const startColumn = column;

    if (ch === '(') {
      tokens.push({
        type: TokenType.LPAREN,
        value: '(',
        line: startLine,
        column: startColumn,
      });
      advance();
    } else if (ch === ')') {
      tokens.push({
        type: TokenType.RPAREN,
        value: ')',
        line: startLine,
        column: startColumn,
      });
      advance();
    } else if (ch === '{') {
      tokens.push({
        type: TokenType.LBRACE,
        value: '{',
        line: startLine,
        column: startColumn,
      });
      advance();
    } else if (ch === '}') {
      tokens.push({
        type: TokenType.RBRACE,
        value: '}',
        line: startLine,
        column: startColumn,
      });
      advance();
    } else if (ch === '[') {
      tokens.push({
        type: TokenType.LBRACKET,
        value: '[',
        line: startLine,
        column: startColumn,
      });
      advance();
    } else if (ch === ']') {
      tokens.push({
        type: TokenType.RBRACKET,
        value: ']',
        line: startLine,
        column: startColumn,
      });
      advance();
    } else if (ch === ':') {
      tokens.push({
        type: TokenType.COLON,
        value: ':',
        line: startLine,
        column: startColumn,
      });
      advance();
    } else if (ch === ',') {
      tokens.push({
        type: TokenType.COMMA,
        value: ',',
        line: startLine,
        column: startColumn,
      });
      advance();
    } else if (ch === '\n') {
      tokens.push({
        type: TokenType.NEWLINE,
        value: '\n',
        line: startLine,
        column: startColumn,
      });
      advance();
    } else if (ch === '"' || ch === "'") {
      const value = readQuotedString(ch);
      tokens.push({
        type: TokenType.QUOTED_STRING,
        value,
        line: startLine,
        column: startColumn,
      });
    } else if (ch && /[a-zA-Z0-9_]/.test(ch)) {
      const value = readBareWord();

      // Check for array marker
      if (current() === '[' && peek() === ']') {
        tokens.push({
          type: TokenType.LABEL_OR_VALUE,
          value,
          line: startLine,
          column: startColumn,
        });
        advance();
        advance();
        tokens.push({
          type: TokenType.ARRAY_MARKER,
          value: '[]',
          line,
          column: column - 2,
        });
      } else {
        tokens.push({
          type: TokenType.LABEL_OR_VALUE,
          value,
          line: startLine,
          column: startColumn,
        });
      }
    } else {
      // Unknown character, try to read as unquoted value
      const value = readUnquotedValue();
      if (value) {
        tokens.push({
          type: TokenType.LABEL_OR_VALUE,
          value,
          line: startLine,
          column: startColumn,
        });
      }
    }
  }

  tokens.push({
    type: TokenType.EOF,
    value: '',
    line,
    column,
  });

  return tokens;
}

File: packages/lmon-python/lmon/tokenizer.py
from enum import Enum
from dataclasses import dataclass
from typing import List, Optional


class TokenType(Enum):
    """Token types for LMON lexical analysis."""
    LPAREN = "LPAREN"
    RPAREN = "RPAREN"
    LBRACE = "LBRACE"
    RBRACE = "RBRACE"
    LBRACKET = "LBRACKET"
    RBRACKET = "RBRACKET"
    COLON = "COLON"
    COMMA = "COMMA"
    LABEL_OR_VALUE = "LABEL_OR_VALUE"
    QUOTED_STRING = "QUOTED_STRING"
    ARRAY_MARKER = "ARRAY_MARKER"
    NEWLINE = "NEWLINE"
    EOF = "EOF"


@dataclass
class Token:
    """A lexical token."""
    type: TokenType
    value: str
    line: int
    column: int


def tokenize(input_text: str) -> List[Token]:
    """Tokenize LMON text into a list of tokens."""
    tokens: List[Token] = []
    i = 0
    line = 1
    column = 1

    def current() -> Optional[str]:
        return input_text[i] if i < len(input_text) else None

    def peek(offset: int = 1) -> Optional[str]:
        pos = i + offset
        return input_text[pos] if pos < len(input_text) else None

    def advance() -> None:
        nonlocal i, line, column
        if i < len(input_text):
            if input_text[i] == '\n':
                line += 1
                column = 1
            else:
                column += 1
            i += 1

    def skip_whitespace() -> None:
        while current() and current() in ' \t\r':
            advance()

    def read_quoted_string(quote: str) -> str:
        nonlocal i, line, column
        advance()  # skip opening quote
        value = ''
        while current() and current() != quote:
            if current() == '\\':
                advance()
                escaped = current()
                if escaped == 'n':
                    value += '\n'
                elif escaped == 't':
                    value += '\t'
                elif escaped == '\\':
                    value += '\\'
                elif escaped == quote:
                    value += quote
                else:
                    value += escaped or ''
                advance()
            else:
                value += current() or ''
                advance()
        if current() == quote:
            advance()
        return value

    def read_bare_word() -> str:
        nonlocal i
        value = ''
        while current() and current() in 'abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789_':
            value += current()
            advance()
        return value

    def read_unquoted_value() -> str:
        nonlocal i
        value = ''
        while (current() and
               current() not in ',{}[]\n' and
               current().strip()):  # not just whitespace
            if current() == '\\':
                advance()
                escaped = current()
                if escaped == 'n':
                    value += '\n'
                elif escaped == 't':
                    value += '\t'
                elif escaped == '\\':
                    value += '\\'
                elif escaped in ',:{}[]()':
                    value += escaped or ''
                else:
                    value += escaped or ''
                advance()
            else:
                value += current() or ''
                advance()
        return value.rstrip()

    while i < len(input_text):
        skip_whitespace()

        if i >= len(input_text):
            break

        ch = current()
        start_line = line
        start_column = column

        if ch == '(':
            tokens.append(Token(TokenType.LPAREN, '(', start_line, start_column))
            advance()
        elif ch == ')':
            tokens.append(Token(TokenType.RPAREN, ')', start_line, start_column))
            advance()
        elif ch == '{':
            tokens.append(Token(TokenType.LBRACE, '{', start_line, start_column))
            advance()
        elif ch == '}':
            tokens.append(Token(TokenType.RBRACE, '}', start_line, start_column))
            advance()
        elif ch == '[':
            tokens.append(Token(TokenType.LBRACKET, '[', start_line, start_column))
            advance()
        elif ch == ']':
            tokens.append(Token(TokenType.RBRACKET, ']', start_line, start_column))
            advance()
        elif ch == ':':
            tokens.append(Token(TokenType.COLON, ':', start_line, start_column))
            advance()
        elif ch == ',':
            tokens.append(Token(TokenType.COMMA, ',', start_line, start_column))
            advance()
        elif ch == '\n':
            tokens.append(Token(TokenType.NEWLINE, '\n', start_line, start_column))
            advance()
        elif ch in '"\'':
            value = read_quoted_string(ch)
            tokens.append(Token(TokenType.QUOTED_STRING, value, start_line, start_column))
        elif ch in 'abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789_':
            value = read_bare_word()

            # Check for array marker
            if current() == '[' and peek() == ']':
                tokens.append(Token(TokenType.LABEL_OR_VALUE, value, start_line, start_column))
                advance()
                advance()
                tokens.append(Token(TokenType.ARRAY_MARKER, '[]', line, column - 2))
            else:
                tokens.append(Token(TokenType.LABEL_OR_VALUE, value, start_line, start_column))
        else:
            value = read_unquoted_value()
            if value:
                tokens.append(Token(TokenType.LABEL_OR_VALUE, value, start_line, start_column))

    tokens.append(Token(TokenType.EOF, '', line, column))
    return tokens
